extensions/notebook/src/extension.ts

```diff
@@ -231,6 +251,87 @@ async function addCell(cellType: azdata.nb.CellType): Promise<void> {
+							title = markdownUrl.substring(markdownUrl.lastIndexOf(path.sep) + 1).replace(/(\r\n|\n|\r)/gm, '');
+							replacedString = `\n- title: ${title}\n  url: /${title}\n  not_numbered: true\n  expand_sections: true\n  sections:\n` + url;
+							result = result.replace(markdownUrl, '');
+						} else {
+							replacedString = `\n- title: ${title}\n  url: /${title}\n  not_numbered: true\n`;
+						}
+						result = result.replace(url, replacedString);
+					});
+				}
+				if (headers) {
+					headers.map(header => {
+						title = header.substring(10).replace(/(\r\n|\n|\r)/gm, '');
+						// for the url you need to have home page on clicking on the header section
+						// check urls that are inside this folder and pick the first one as home
+						contentFolders.push(title.toLocaleLowerCase());
+						let filtered: string[] = urls.filter(x => x.indexOf(`${path.sep}${title.toLocaleLowerCase()}${path.sep}`) > -1);
+						// toc has markdown files listed at the bottom.
+						let index: number = urls.findIndex(x => x === filtered[filtered.length - 1]);
+						let urlValue: string = urls[index].substring(urls[index].lastIndexOf(path.sep) + 1).replace(/(\r\n|\n|\r)/gm, '');
+						replacedString = `- title: ${title}\n  url: /${title}/${urlValue}\n  not_numbered: true\n  expand_sections: true\n  sections:\n`;
+						result = result.replace(header, replacedString);
+						result = result.replace(urls[index], '');
+						urls.splice(index, 1);
+					});
+				}
+				if (urls) {
+					urls.map(url => {
+						title = url.substring(url.lastIndexOf(path.sep) + 1).replace(/(\r\n|\n|\r)/gm, '');
+						let urlValue: string = title;
+						// if there are folders inside content, the url will be that foldrname/filename
+						if (contentFolders.length > 0) {
+							// get each folder from the path bread crumb and check if the last folder is inside content
+							let folders: string[] = url.substring(8).split(path.sep);
+							let parentFolder: number = contentFolders.findIndex(x => x === folders[folders.length - 2]);
+							if (parentFolder > -1) {
+								urlValue = `${contentFolders[parentFolder]}/${title}`;
+							}
+						}
+						replacedString = `  - title: ${title}\n    url: /${urlValue}\n`;
+						result = result.replace(url, replacedString);
+					});
+				}
+				fsPromises.writeFile(tocFilePath, result);
+			} else {
+				new ApiWrapper().showErrorMessage('File Name contains unsupported-characters (ex: underscores or spaces) by Jupyter Book');
+			}
+		});
+		// update the book title
+		let configFilePath: string = path.join(bookPath, '_config.yml');
+		fsPromises.readFile(configFilePath).then(data => {
+			let titleLine = data.toString().match(/title: [a-zA-Z0-9\\.\s-]+$/gm);
+			let title: string = `title: ${path.basename(bookPath)}`;
+			result = data.toString().replace(titleLine[0], title);
+			fsPromises.writeFile(configFilePath, result);
+		});
+	} catch (error) {
+		new ApiWrapper().showErrorMessage(error);
```

Comment:
This shouldn't be recreated each time - just create it once in the activate function and pass it through to this and the AppContext as needed